### PR TITLE
Ping on cluster using Predis

### PR DESCRIPTION
**Pull request by jack9603301**
Redis Cluster does not support the Ping command

### CODE REVIEW

**Comment by tillkruss** [comment]
I haven't gotten any reports on this. Why was it merged in #70? 

What client and version are you using?

**Comment by jack9603301** [comment]
I use the latest redis cluster and predis library to link. I don't know why I was contracted in #70, but in my environment, the ping command can't be submitted to the cluster by predis!

**Comment by tillkruss** [comment]
I won’t be supporting Predis much longer. Can you please test this using PhpRedis.

**Comment by jack9603301** [comment]
The following error occurred after I changed to PECL:
```
Couldn't map cluster keyspace using any provided seed (RedisClusterException)
```

**Comment by tillkruss** [comment]
It looks like you're trying to use cluster mode, but your running a regular Redis instace.

Does using `redis-cli -c -p 6380` work as expected?

**Comment by jack9603301** [comment]
My redis cluster uses the following command to access normally:

```
redis-cli -a password -c -p port 
```

**Comment by tillkruss** [comment]
Ping is working with PhpRedis, feel free to update this PR only disabling ping on clusters when Predis is used.

**Comment by jack9603301** [comment]
But I already did what you said in this PR, please let me know if I make any mistake on this PR, otherwise, you can merge this PR.

**Comment by jack9603301** [comment]
please let me know if I make any mistake on this PR, otherwise, you can merge this PR.

File: includes/object-cache.php [hunk 1]
```diff
@@ -566,13 +566,14 @@ public function __construct($fail_gracefully = true)
                 $this->redis_client .= sprintf(' (v%s)', Predis\Client::VERSION);
             }
 
-            if (defined('WP_REDIS_CLUSTER')) {
-                $this->redis->ping(current(array_values(WP_REDIS_CLUSTER)));
+            if ((strcasecmp('predis', $client) === 0) && defined('WP_REDIS_CLUSTER')) {
+                // Redis Cluster does not support the Ping command
+                //$this->redis->ping(current(array_values(WP_REDIS_CLUSTER)));
```
Comment (tillkruss) [inline includes/object-cache.php]: This will break pings for non Predis clusters.

File: includes/object-cache.php [hunk 2]
```diff
@@ -566,13 +566,14 @@ public function __construct($fail_gracefully = true)
...
             } else {
                 $this->redis->ping();
             }
 
-            if ( ! isset( $options['replication'] ) || ! $options['replication'] ) {
+            if ( ((strcasecmp('predis', $client) === 0) && !defined('WP_REDIS_CLUSTER')) && (! isset( $options['replication'] ) || ! $options['replication']) ) {
```
Comment (tillkruss) [inline includes/object-cache.php]: These should be extracted into variables, such as `$usesReplication` and `$onPredisCluster`.

File: includes/diagnostics.php [hunk 3]
```diff
@@ -6,15 +6,24 @@
...
+}
 $info[ 'Client' ] = $this->get_redis_client_name();
 
 $info[ 'Drop-in' ] = $dropin ? 'Valid' : 'Invalid';
 
 if ( $dropin ) {
     try {
-        $cache = new WP_Object_Cache( false );
-        $info[ 'Ping' ] = defined( 'WP_REDIS_CLUSTER' ) ? 'Not supported' : $cache->redis_instance()->ping();
+        if (defined( 'WP_REDIS_CLIENT' ) && WP_REDIS_CLIENT === 'predis' && defined( 'WP_REDIS_CLUSTER' )) {
```
Comment (tillkruss) [inline includes/diagnostics.php]: This is redundant. Should be extracted into `$usesPredis` boolean.

File: includes/diagnostics.php [hunk 4]
```diff
@@ -6,15 +6,24 @@
 $dropin = $this->validate_object_cache_dropin() && ( ! defined('WP_REDIS_DISABLED') || ! WP_REDIS_DISABLED );
 
 $info[ 'Status' ] = $this->get_status();
-$info[ 'Redis Version' ] = $this->get_redis_version() ?: 'Unknown';
+
+if (defined( 'WP_REDIS_CLIENT' ) && WP_REDIS_CLIENT === 'predis' && defined( 'WP_REDIS_CLUSTER' )) {
+    $info[ 'Redis Version' ] = 'Not supported';
+} else {
+    $info[ 'Redis Version' ] = $this->get_redis_version() ?: 'Unknown';
```
Comment (tillkruss) [inline includes/diagnostics.php]: Wouldn't this just default to `Unknown`? Is this condition really necessary.

Comment (jack9603301) [reply includes/diagnostics.php]: Original judgement
```
$info[ 'Redis Version' ] = $this->get_redis_version() ?: 'Unknown';
```
It may not work, so I made a change, in fact, it does have problems in my environment! And not in line with the general way of judgment! The format of ternary logic judge is:
```
boolean?true:false
```
So the part of true should not be empty

**Comment by jack9603301** [comment]
Please continue to review and feedback the existing code, otherwise, it can be merged!

File: includes/diagnostics.php [hunk 5]
```diff
@@ -6,15 +6,26 @@
 $dropin = $this->validate_object_cache_dropin() && ( ! defined('WP_REDIS_DISABLED') || ! WP_REDIS_DISABLED );
 
 $info[ 'Status' ] = $this->get_status();
-$info[ 'Redis Version' ] = $this->get_redis_version() ?: 'Unknown';
+
+if (defined( 'WP_REDIS_CLIENT' ) && WP_REDIS_CLIENT === 'predis' && defined( 'WP_REDIS_CLUSTER' )) {
+    $info[ 'Redis Version' ] = 'Not supported';
+} else {
+    $info[ 'Redis Version' ] = $this->get_redis_version();
+}
```
Comment (tillkruss) [inline includes/diagnostics.php]: Why is this necessary? What is the exact error you're seeing without it?

Comment (jack9603301) [reply includes/diagnostics.php]: 1. The preferred condition to check whether version is available should not be determined by whether the normal value of `$this->get_redis_version()` is true or false
2. Boolean expression in the form: 
```
Boolean? True: false
```

as well as

```
$this->get_redis_version()?:"Unkwon";
```

true or false should be returned for `$this->get_redis_version()`, not other return types, and should avoid the possibility of throwing exceptions!

3. For a Boolean expression, the execution logic is that when Boolean returns a Boolean value, the code of true or false segment and its return value are executed and returned according to the Boolean value. Therefore, the true segment should not be empty!

4. Maybe my writing method is not the only one, but it can guarantee that there will be no serious problems in logic! In fact, a conforming Boolean expression should have the same effect on its equivalent code. Any conforming Boolean expression can be converted to the form of `if-else-endif` normally。

When I modified the bug, I found that redis cluster was unable to execute the server command to obtain version information. After rewriting the judgment during debugging and running normally, I decided to modify the judgment logic of the location based on the above reasons!

Comment (tillkruss) [reply includes/diagnostics.php]: It's safe cast strings as booleans: https://www.php.net/manual/en/types.comparisons.php

What's the exception you are seeing without your fix?

Comment (jack9603301) [reply includes/diagnostics.php]: It is not a matter of type conversion, but of the meaning of Boolean expressions, such as:
```
$tmp = fun() ? 1:2
```
The fun function will be called. If the return value of the function is true, the expression will return 1, otherwise it will return 2. In the original written expression, there is no true condition specified for the Boolean expression. Moreover, if the Boolean condition of the expression is function function call, once the function throws an exception, it will not execute normally!

It's been several days since the code was modified. If I want to reproduce it, I have to modify it again. It's troublesome, so I won't reproduce it. You can try it yourself! Since the later modified code can run normally and meet the specification, it is recommended to keep it!

Comment (tillkruss) [reply includes/diagnostics.php]: Why would `get_redis_version()` ever throw an exception?

**Comment by jack9603301** [comment]
In fact, I don't think the key is to throw an exception, because the original writing method is:

```
$info[ 'Redis Version' ] = $this->get_redis_version() ?: 'Unknown';
```

If `$this - > get_redis_version() ` returns true, the expression will have no return value!

**Comment by tillkruss** [comment]
> If `$this - > get_redis_version() ` returns true, the expression will have no return value!

Exactly, unless a truthy value is returned, that expression will show `Unknown`, that's you can undo that change 👍🏻

**Comment by jack9603301** [comment]
I have modified the logic and reduced it to the equivalent writing method of logical expression. Can you see?

File: includes/object-cache.php [hunk 6]
```diff
@@ -566,13 +566,20 @@ public function __construct($fail_gracefully = true)
...
+                }
             } else {
                 $this->redis->ping();
             }
 
-            if ( ! isset( $options['replication'] ) || ! $options['replication'] ) {
+            $usesReplication = isset($options['replication']) && $options['replication'];
+
+            if (! $usesReplication) {
                 $server_info = $this->redis->info( 'SERVER' );
```
Comment (tillkruss) [inline includes/object-cache.php]: @jack9603301: Can you please test this code with your Predis cluster and post the exception it throws, if it throws any?

Comment (jack9603301) [reply includes/object-cache.php]: There is no judgment. I have submitted relevant corrections. Please see what I submitted. The reason why I made a judgment is that I found that the `server` instruction may have execution errors in predis mode. It is forbidden to execute the command for redis cluster.

Comment (tillkruss) [reply includes/object-cache.php]: I will not merge this pull request, unless I understand it and see the corresponding exception/error message.

Comment (jack9603301) [reply includes/object-cache.php]: I remember that I gave you an exception about ping command execution error, which is the same as the exception!

Comment (jack9603301) [reply includes/object-cache.php]: I remember that I gave you an exception about ping command execution error, which is the same as the exception!

**Comment by tillkruss** [comment]
I don't feel comfortable merging this pull request until exception messages for all code changes are provided.
